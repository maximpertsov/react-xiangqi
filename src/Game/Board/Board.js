/** @jsx jsx */
import { jsx, css } from '@emotion/core';

import { useState, useLayoutEffect } from 'react';
import PropTypes from 'prop-types';
import Square from '../Square/Square';
import { boardPropType } from '../../logic';
import * as styles from '../../commonStyles';

import boardImg from './board-1000px.svg.png';

const STEP_TINY = '20px';
const STEP_SMALL = '30px';
const STEP_MEDIUM = '45px';
const STEP_LARGE = '60px';

const Board = ({
  board,
  handleLegalMove,
  legalMoves,
  nextMoveColor,
  reversed,
}) => {
  const [selectedSlot, setSelectedSlot] = useState(undefined);
  const [moveX, setMoveX] = useState(0);
  const [moveY, setMoveY] = useState(0);

  // NOTE: this is the synchronous version of useEffect. Using
  // this version prevents late selection clearing, but causes
  // move updates to take longer. Annoying in development but
  // seems to be faster enough on a production build.
  useLayoutEffect(
    () => { setSelectedSlot(undefined); },
    // TODO: is it too expensive to check if the board changes?
    // Can I key on another prop update?
    [board],
  );

  const getPieceCode = (slot) => board.getPiece(slot) || undefined;

  const selectedCanCapture = (slot) => {
    if (selectedSlot === undefined) return false;
    if (!board.isOccupied(selectedSlot)) return false;
    if (!board.isOccupied(slot)) return false;
    return !board.sameColor(slot, selectedSlot);
  };

  const isLegalMove = (fromSlot, toSlot) => (
    legalMoves[fromSlot].includes(toSlot)
  );

  const handleMove = async(fromSlot, toSlot) => {
    if (isLegalMove(fromSlot, toSlot)) {
      const [fromY, fromX] = board.getRankFile(fromSlot);
      const [toY, toX] = board.getRankFile(toSlot);
      setMoveX(toX - fromX);
      setMoveY(toY - fromY);
      setTimeout(() => {
        handleLegalMove({ board, fromSlot, toSlot });
        setMoveX(0);
        setMoveY(0);
      }, 150);
    } else {
      setSelectedSlot(undefined);
    }
  };

  const handleSquareClick = (slot) => (() => {
    if (slot === selectedSlot) {
      setSelectedSlot(undefined);
    } else if (board.isOccupied(slot) && !selectedCanCapture(slot)) {
      setSelectedSlot(slot);
    } else if (selectedSlot !== undefined) {
      handleMove(selectedSlot, slot);
    } else {
      setSelectedSlot(undefined);
    }
  });

  const getTargets = () => (
    selectedSlot === undefined ? [] : legalMoves[selectedSlot]
  );

  const inCheck = (slot) => {
    if (!board.kingInCheck(nextMoveColor)) return false;
    return board.findKingSlot(nextMoveColor) === slot;
  };

  const getSlot = (b, i) => (reversed ? b.length - i - 1 : i);

  const renderSquares = () => board.board.map((_, i, b) => {
    const slot = getSlot(b, i);
    return (
      <Square
        key={slot}
        handleClick={handleSquareClick(slot)}
        inCheck={inCheck(slot)}
        pieceCode={getPieceCode(slot)}
        selected={selectedSlot === slot}
        targeted={getTargets().includes(slot)}
        moveX={moveX}
        moveY={moveY}
      />
    );
  });

  return (
    <div
      className="Board"
      css={css`
        background-image: url(${boardImg});
        background-size: contain;
        background-repeat: no-repeat;
        background-position: top;
        display: grid;
        ${styles.MEDIA_TINY} {
          grid-template-rows: repeat(10, ${STEP_TINY});
          grid-template-columns: repeat(9, ${STEP_TINY});
        }
        ${styles.MEDIA_SMALL} {
          grid-template-rows: repeat(10, ${STEP_SMALL});
          grid-template-columns: repeat(9, ${STEP_SMALL});
        }
        ${styles.MEDIA_MEDIUM} {
          grid-template-rows: repeat(10, ${STEP_MEDIUM});
          grid-template-columns: repeat(9, ${STEP_MEDIUM});
        }
        ${styles.MEDIA_LARGE} {
          grid-template-rows: repeat(10, ${STEP_LARGE});
          grid-template-columns: repeat(9, ${STEP_LARGE});
        }
      `}
    >
      {renderSquares()}
    </div>
  );
};

Board.propTypes = {
  board: boardPropType.isRequired,
  handleLegalMove: PropTypes.func.isRequired,
  legalMoves: PropTypes.arrayOf(PropTypes.arrayOf(PropTypes.number)).isRequired,
  nextMoveColor: PropTypes.string.isRequired,
  reversed: PropTypes.bool.isRequired,
};

export default Board;
