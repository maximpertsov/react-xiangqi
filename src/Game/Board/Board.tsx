/** @jsx jsx */ jsx

import React from 'react'
import styled from '@emotion/styled'
import PropTypes from 'prop-types'
import Square from '../Square/Square'
import { css, jsx } from '@emotion/core'
import { getPiece } from '../Piece/Piece'
import { boardPropType } from '../../logic'

import boardImg from './board-1000px.svg.png'

// TODO find a better name
type XBoard = any
// TODO move to central location
type Slot = number | 

interface BoardProps {
  board: XBoard
  handleLegalMove: (board: XBoard, fromSlot: number, toSlot: number) => void
  handleSelect: (params: { slot: number }) => void
  legalMoves: number[][]
  nextMoveColor: number
  reversed: boolean
  selectedSlot?: number | null
}

const Board = ({
  board,
  handleLegalMove,
  handleSelect,
  legalMoves,
  nextMoveColor,
  reversed,
  selectedSlot = null,
}: BoardProps) => {
  const getPieceOn = (slot: number) => getPiece(board.getPiece(slot))

  const selectedCanCapture = (slot: number) => {
    if (selectedSlot === null) return false
    if (!board.isOccupied(selectedSlot)) return false
    if (!board.isOccupied(slot)) return false
    return !board.sameColor(slot, selectedSlot)
  }

  const isLegalMove = (fromSlot: number, toSlot: number) =>
    legalMoves[fromSlot].includes(toSlot)

  const handleMove = (fromSlot: number, toSlot: number) => {
    if (isLegalMove(fromSlot, toSlot)) handleLegalMove(board, fromSlot, toSlot)
    handleSelect({ slot: null })
  }

  const handleSquareClick = ({ slot: number, isOccupied: boolean }) => {
    if (slot === selectedSlot) {
      handleSelect({ slot: null })
    } else if (isOccupied && !selectedCanCapture(slot)) {
      handleSelect({ slot })
    } else if (selectedSlot !== null) {
      handleMove(selectedSlot, slot)
    } else {
      handleSelect({ slot: null })
    }
  }

  const getTargets = () =>
    selectedSlot === null ? [] : legalMoves[selectedSlot]

  const getInCheckSlot = () => {
    if (!board.kingInCheck(nextMoveColor)) return undefined
    return board.findKingSlot(nextMoveColor)
  }

  const getSlot = (b, i) => (reversed ? b.length - i - 1 : i)

  const renderSquares = () =>
    board.board.map((_, i, b) => {
      const slot = getSlot(b, i)
      return (
        <Square
          key={slot}
          slot={slot}
          piece={getPieceOn(slot)}
          inCheckSlot={getInCheckSlot()}
          targets={getTargets()}
          handleSquareClick={handleSquareClick}
          selected={selectedSlot === slot}
        />
      )
    })

  return (
    <div
      className="Board"
      css={css`
        background-image: url(${boardImg});
        background-size: contain;
        background-repeat: no-repeat;
        background-position: top;
        display: grid;
        grid-template-rows: repeat(10, 60px);
        grid-template-columns: repeat(9, 60px);
        justify-content: center;
      `}
    >
      {renderSquares()}
    </div>
  )
}
export default Board
