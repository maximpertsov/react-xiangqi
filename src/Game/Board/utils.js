export const cellID = (i, j) => `square_${i}_${j}`;

export default {};
