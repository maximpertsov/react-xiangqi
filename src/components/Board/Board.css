.Board {
  display: grid;
  grid-template-rows: repeat(10, 60px);
  grid-template-columns: repeat(9, 60px);
  justify-content: center;
}

img {
	display: block;
  margin-left: auto;
  margin-right: auto;
	max-width: 100%;
}
