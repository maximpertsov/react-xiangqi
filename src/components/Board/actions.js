export const setSelectedSlot = ({ selectedSlot }) => ({
  type: 'set_selected_slot',
  selectedSlot,
});

export const clearSelectedSlot = () => setSelectedSlot({ selectedSlot: null });

export const setAnimationOffset = ({
  bottomPlayerIsRed,
  fromSlot,
  toSlot,
}) => ({
  type: 'set_animation_offset',
  bottomPlayerIsRed,
  fromSlot,
  toSlot,
});

export const clearAnimationOffset = () => ({
  type: 'clear_animation_offset',
});

export default {};
