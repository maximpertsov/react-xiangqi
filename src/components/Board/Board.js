import React, { Component } from 'react';
import './Board.css';

class Board extends Component {
  render() {
    return (
      <div className="Board">
        <img src="https://upload.wikimedia.org/wikipedia/commons/6/66/Xiangqi-_md.PNG" />
        <img src="https://upload.wikimedia.org/wikipedia/commons/6/66/Xiangqi-_md.PNG" />
        <img src="https://upload.wikimedia.org/wikipedia/commons/6/66/Xiangqi-_md.PNG" />
        <img src="https://upload.wikimedia.org/wikipedia/commons/6/66/Xiangqi-_md.PNG" />
        <img src="https://upload.wikimedia.org/wikipedia/commons/6/66/Xiangqi-_md.PNG" />
        <img src="https://upload.wikimedia.org/wikipedia/commons/6/66/Xiangqi-_md.PNG" />
        <img src="https://upload.wikimedia.org/wikipedia/commons/6/66/Xiangqi-_md.PNG" />
        <img src="https://upload.wikimedia.org/wikipedia/commons/6/66/Xiangqi-_md.PNG" />
        <img src="https://upload.wikimedia.org/wikipedia/commons/6/66/Xiangqi-_md.PNG" />
        <img src="https://upload.wikimedia.org/wikipedia/commons/6/66/Xiangqi-_md.PNG" />
        <img src="https://upload.wikimedia.org/wikipedia/commons/6/66/Xiangqi-_md.PNG" />
        <img src="https://upload.wikimedia.org/wikipedia/commons/6/66/Xiangqi-_md.PNG" />
        <img src="https://upload.wikimedia.org/wikipedia/commons/6/66/Xiangqi-_md.PNG" />
        <img src="https://upload.wikimedia.org/wikipedia/commons/6/66/Xiangqi-_md.PNG" />
        <img src="https://upload.wikimedia.org/wikipedia/commons/6/66/Xiangqi-_md.PNG" />
      </div>
    );
  }
}

export default Board;
