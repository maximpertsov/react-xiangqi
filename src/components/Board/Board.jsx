import React, { Component } from 'react';
import styled from '@emotion/styled';
import Square from '../Square/Square';
import layout from '../Piece/utils';
import { cellID } from './utils';

import boardImg from './board-1000px.svg.png';

const Wrapper = styled.div`
  background-image: url(${boardImg});
  background-size: contain;
  background-repeat: no-repeat;
  background-position: center;
  display: grid;
  grid-template-rows: repeat(10, 60px);
  grid-template-columns: repeat(9, 60px);
  justify-content: center;
`;

class Board extends Component {
  constructor(props) {
    super(props);

    this.handleClick = this.handleClick.bind(this);

    this.state = {
      pieces: layout,
      selected: null,
    };
  }

  handleClick(cellName) {
    this.setState({ selected: cellName });
  }

  render() {
    const { pieces, selected } = this.state;

    return (
      <Wrapper className="Board">
        {pieces.map((row, i) => (
          row.map((p, j) => (
            <Square
              key={cellID(i, j)}
              cellName={cellID(i, j)}
              piece={p}
              selected={selected === cellID(i, j)}
              handleClick={this.handleClick}
            />
          ))
        ))}
      </Wrapper>
    );
  }
}

export default Board;
