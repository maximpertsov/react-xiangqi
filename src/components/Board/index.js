import React, { useEffect } from 'react';
import { DndProvider } from 'react-dnd';
import Backend from 'react-dnd-html5-backend';
import { useDispatch, useSelector } from 'react-redux';
import { createSelector } from 'reselect';
import get from 'lodash/get';
import isEqual from 'lodash/isEqual';

import makeMove from 'actions/makeMove';
import actions from 'actions';
import animateMove from 'actions/animateMove';
import { getBottomPlayerIsRed, getLegalMoves } from 'reducers';
import { isOccupied, sameTeam } from 'services/logic/fen';
import { squaresToUci } from 'services/logic/square';

import BoardView from './components/BoardView';

const ANIMATION_DELAY = 150;

const mapStateToProps = createSelector(
  [state => state],

  state => ({
    bottomPlayerIsRed: getBottomPlayerIsRed(state),
    legalMoves: getLegalMoves(state),
    selectedFen: state.selectedFen,
    selectedSquare: state.selectedSquare,
  }),
);

const Board = () => {
  const dispatch = useDispatch();

  const {
    bottomPlayerIsRed,
    legalMoves,
    selectedFen,
    selectedSquare,
  } = useSelector(mapStateToProps, isEqual);

  useEffect(() => {
    dispatch(actions.board.selectedSquare.set(null));
  }, [dispatch, selectedFen]);

  const selectedCanCapture = square => {
    if (selectedSquare === null) return false;
    if (!isOccupied(selectedFen, selectedSquare)) return false;
    if (!isOccupied(selectedFen, square)) return false;

    return !sameTeam(selectedFen, square, selectedSquare);
  };

  const legalFen = uci => get(legalMoves, uci, false);

  const handleMove = uci => {
    const fen = legalFen(uci);
    if (fen) {
      dispatch(animateMove({ bottomPlayerIsRed, uci }));
      setTimeout(() => {
        dispatch(makeMove({ fen, uci }));
        dispatch(actions.board.animationOffset.clear());
        dispatch(actions.board.selectedSquare.set(null));
      }, ANIMATION_DELAY);
    } else {
      dispatch(actions.board.selectedSquare.set(null));
    }
  };

  const handleSquareClick = square => () => {
    if (square === selectedSquare) {
      dispatch(actions.board.selectedSquare.set(null));
    } else if (isOccupied(selectedFen, square) && !selectedCanCapture(square)) {
      dispatch(actions.board.selectedSquare.set(square));
    } else if (selectedSquare !== null) {
      handleMove(squaresToUci(selectedSquare, square));
    } else {
      dispatch(actions.board.selectedSquare.set(null));
    }
  };

  return (
    <DndProvider backend={Backend}>
      <BoardView handleSquareClick={handleSquareClick} />
    </DndProvider>
  );
};

export default Board;
