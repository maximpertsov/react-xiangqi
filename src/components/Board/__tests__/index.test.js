import React from 'react';
import { useSelector } from 'react-redux';

import Board from '..';

jest.mock('react-redux');
jest.mock('reducers/selectors');

describe('Board', () => {
  const move = {
    fen:
      'rnbak1bnr/4a4/1c5c1/p1p1p1p1p/9/2P3P2/P3P3P/1C5C1/9/RNBAKABNR b - - 3 2',
    givesCheck: false,
    uci: 'g4g5',
  };
  let store;
  let wrapper;

  beforeEach(() => {
    store = mockStore({});
    useSelector.mockImplementation(callback => callback(store.getState()));
  });

  afterEach(() => {
    jest.resetAllMocks();
  });

  describe('bottom player has red pieces', () => {
    beforeEach(() => {
      wrapper = shallowWrappedComponent(<Board move={move} />, store);
    });

    test('snapshot', () => {
      expect(wrapper).toMatchSnapshot();
    });
  });

  describe('bottom player has black pieces', () => {
    beforeEach(() => {
      wrapper = shallowWrappedComponent(
        <Board teamBlackPOV move={move} />,
        store,
      );
    });

    test('snapshot', () => {
      expect(wrapper).toMatchSnapshot();
    });
  });
});
