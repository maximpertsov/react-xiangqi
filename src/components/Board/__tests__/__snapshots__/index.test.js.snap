// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Board snapshot 1`] = `
<BoardView
  handleSquareClick={[Function]}
  move={Object {}}
  size="fluid"
  teamBlackPOV={false}
/>
`;
